### PR TITLE
large refactoring

### PR DESCRIPTION
**Pull request by jaredatron**

### CODE REVIEW

**Comment by jaredatron** [comment]
@smtlaissezfaire @pariser check it :D

File: Gemfile [hunk 1]
```diff
@@ -1,9 +1,11 @@
 source 'https://rubygems.org'
 
 gem 'mongo_mapper', '0.12.0'
+gem 'mongo', '1.11.1'
```
Comment (smtlaissezfaire) [inline Gemfile:4]: Why do you want to lock mongo version?

File: Gemfile [hunk 2]
```diff
@@ -1,9 +1,11 @@
...
+gem 'plucky', '0.5.2'
```
Comment (smtlaissezfaire) [inline Gemfile:5]: plucky is a requirement of mongo_mapper.  I don't think you want to lock it, do you?

File: Gemfile [hunk 3]
```diff
@@ -1,9 +1,11 @@
...
 
 group :test do
-  gem 'rspec'
+  gem 'rspec', '2.14.1'
```
Comment (smtlaissezfaire) [inline Gemfile:8]: why lock rspec?

File: Gemfile (hunk 2 above)
Comment (jaredatron) [inline Gemfile:5]: This doesn't "lock it" anymore than Bundler ever locks it. It just prevents the gem developer from upgrading away from this version unintentionally.

File: Gemfile (hunk 3 above)
Comment (jaredatron) [inline Gemfile:8]: again, not locked just specified

File: Gemfile (hunk 1 above)
Comment (jaredatron) [inline Gemfile:4]: It's the version we use so I wanted to ensure that in development of this gem were developing against the same version

File: lib/mongo_mapper/eager_includer.rb [hunk 4]
```diff
@@ -1,194 +1,165 @@
...
-    elsif @proxy_class == MongoMapper::Plugins::Associations::BelongsToProxy
-      eager_include_belongs_to(&block)
-    elsif @proxy_class == MongoMapper::Plugins::Associations::OneProxy
-      eager_include_has_one(&block)
-    elsif @proxy_class == MongoMapper::Plugins::Associations::InArrayProxy
-      eager_include_has_many_in(&block)
+    return if @records.length == 0
+
+    @association_type = case @association.proxy_class.to_s
+    when 'MongoMapper::Plugins::Associations::ManyDocumentsProxy'
```
Comment (smtlaissezfaire) [inline lib/mongo_mapper/eager_includer.rb:33]: Why use strings instead of classes?

File: lib/mongo_mapper/eager_includer.rb [hunk 5]
```diff
@@ -1,194 +1,165 @@
...
+  attr_reader :association_name
 
-    self.class.write_to_cache(record.object_id, association_name, value)
+  def instance_variable_name
+    "@eager_loaded_#{association_name}"
+  end
 
-    code = <<-CODE
+  def setup_association(record, value)
+    code = <<-RUBY
```
Comment (smtlaissezfaire) [inline lib/mongo_mapper/eager_includer.rb:57]: Hmm - we should use:

`<<-RUBY, __FILE__, __LINE__`

here for debuggers / file + line numbers

File: lib/mongo_mapper/eager_includer.rb (hunk 4 above)
Comment (jaredatron) [inline lib/mongo_mapper/eager_includer.rb:33]: I don't know why I made that change.

File: lib/mongo_mapper/eager_includer.rb [hunk 6]
```diff
@@ -1,194 +1,165 @@
...
   end
 
   def foreign_keys
-    @association.options[:in]
+    @foreign_keys ||= @association.options[:in]
   end
 
   def foreign_key
-    @association_name.to_s.foreign_key
+    @foreign_key ||= @association_name.to_s.foreign_key
```
Comment (smtlaissezfaire) [inline lib/mongo_mapper/eager_includer.rb:75]: why cache these?  seems like a minor perf improvement

File: lib/mongo_mapper/eager_includer.rb (hunk 5 above)
Comment (jaredatron) [inline lib/mongo_mapper/eager_includer.rb:57]: `__FILE__, __LINE__` happens later when you eval the code. this just defines a string

File: lib/mongo_mapper/eager_includer.rb (hunk 6 above)
Comment (jaredatron) [inline lib/mongo_mapper/eager_includer.rb:75]: because they dont change?

File: lib/mongo_mapper/eager_includer.rb (hunk 5 above)
Comment (smtlaissezfaire) [inline lib/mongo_mapper/eager_includer.rb:57]: Ah, your correct and it's a few lines later.  N.M.

**Comment by smtlaissezfaire** [comment]
Still reviewing, but I'll merge later today.

FYI - I'm planning on rolling this into MM proper at some point, so remind me to give you credit in the CONTRIBUTORS file.

**Comment by jaredatron** [comment]
yeah! :D

On Mon, May 9, 2016 at 11:32 AM, Scott Taylor notifications@github.com
wrote:

> Still reviewing, but I'll merge later today.
> 
> FYI - I'm planning on rolling this into MM proper at some point, so remind
> me to give you credit in the CONTRIBUTORS file.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> https://github.com/GoLearnUp/mm_eager_includer/pull/1#issuecomment-217949150

## 

Jared Grippe
jared@learnup.com

**Pull request merged**